### PR TITLE
[SPARK-15294][SPARKR][MINOR] Add pivot functionality to SparkR

### PR DESCRIPTION
**Pull request by mhnatiuk**
## What changes were proposed in this pull request?

Added pivot functionality to SparkR
## How was this patch tested?

 R package sucesfully builds and loads

### CODE REVIEW

**Comment by shivaram** [comment]
Jenkins, ok to test

**Comment by shivaram** [comment]
Thanks @mhnatiuk for opening this PR. Could we also add a unit test in `test_sparkSQL.R` for this ?

**Comment by mhnatiuk** [comment]
Jenkins, retest this please

**Comment by mhnatiuk** [comment]
Jenkins, retest this please

**Comment by mhnatiuk** [comment]
Jenkins, retest this please

**Comment by mhnatiuk** [comment]
Jenkins, retest this please

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 1]
```diff
@@ -16,6 +16,7 @@
 #
 
 library(testthat)
+library(magrittr)
```
Comment (shivaram) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I think its better to add a test case that doesn't depend on `magrittr` as our Jenkins environment or other developers might not have the package installed.

File: R/pkg/NAMESPACE [hunk 2]
```diff
@@ -280,7 +280,7 @@ exportMethods("%in%",
               "year")
 
 exportClasses("GroupedData")
-exportMethods("agg")
+exportMethods("agg", "pivot")
```
Comment (felixcheung) [inline R/pkg/NAMESPACE]: nit: i think it would be easier to have one line for each like above

File: R/pkg/R/group.R [hunk 3]
```diff
@@ -126,6 +126,47 @@ methods <- c("avg", "max", "mean", "min", "sum")
 # These are not exposed on GroupedData: "kurtosis", "skewness", "stddev", "stddev_samp", "stddev_pop",
 # "variance", "var_samp", "var_pop"
 
+#' Pivot
+#' 
+#' Pivot GroupedData column to expand unique values of this column into set of column named accordingly and perform
+#' aggregation. Same purpose as dcast in reshape2 or pivot table in Excel
+#' 
+#' @param x: a GroupedData object
+#' @param by: name of column to pivot data by
```
Comment (felixcheung) [inline R/pkg/R/group.R]: this seems to be `colname` below. should this be `colnames`?

File: R/pkg/R/group.R [hunk 4]
```diff
@@ -126,6 +126,47 @@ methods <- c("avg", "max", "mean", "min", "sum")
...
+#' @param values: a unique list of values that will be translated to columns in the output SparkDataFrame. 
+#' If not given it will be computed on the fly which will influence performance
+#' @return GroupedData object
+#' @rdname pivot
+#' @export 
+#' @examples
```
Comment (felixcheung) [inline R/pkg/R/group.R]: please add @family, @name

File: R/pkg/R/group.R [hunk 5]
```diff
@@ -126,6 +126,47 @@ methods <- c("avg", "max", "mean", "min", "sum")
...
+#' values <- list("R", "Python")
+#' sums <- groupBy(SparkRdf, "year") %>% 
+#'     pivot("course", values) %>% 
+#'     SparkR::summarize(sumOfEarnings = sum(SparkRdf$earnings) ) %>%
+#'     collect()
+#' }
+
+setMethod("pivot",
+          signature(x = "GroupedData"),
+          function(x, colname, values=NULL){
```
Comment (felixcheung) [inline R/pkg/R/group.R]: should this be `values=list()`? I think it encourages the use by showing it is a list

**Comment by shivaram** [comment]
Jenkins, add to whitelist

**Comment by shivaram** [comment]
Jenkins, ok to test

File: R/pkg/R/group.R (hunk 5 above)
Comment (mhnatiuk) [inline R/pkg/R/group.R]: OK, so I'll add a test to verify that list ain't empty, right? Or does the underlaying scala method accept empty lists?

**Comment by mhnatiuk** [comment]
Jenkins, retest this please

**Comment by mhnatiuk** [comment]
Jenkins, retest this please

**Comment by mhnatiuk** [comment]
Jenkins, retest this please

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 6]
```diff
@@ -1353,6 +1354,21 @@ test_that("group by, agg functions", {
...
+  df <- data.frame(
+        earnings = c(10000, 10000, 11000, 15000, 12000, 20000, 21000, 22000),
+        course = c("R", "Python", "R", "Python", "R", "Python", "R", "Python"),
+        year = c(2013, 2013, 2014, 2014, 2015, 2015, 2016, 2016)
+  )
+  SparkRdf <- createDataFrame(df)
+  values <- list("R", "Python")
+  sums <- collect(SparkR::summarize(pivot(groupBy(SparkRdf, "year"), "course", values),
+                                    earnings_sum = sum(SparkRdf$earnings)) )
+  testthat::expect_equal(colnames(sums) %in% c("year", "R", "Python"), c(T, T, T))
```
Comment (shivaram) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: we probably dont need `testthat::` here.

File: R/pkg/R/group.R [hunk 7]
```diff
@@ -126,6 +126,49 @@ methods <- c("avg", "max", "mean", "min", "sum")
...
+#'     SparkR::summarize(sumOfEarnings = sum(SparkRdf$earnings) ) %>%
+#'     collect()
+#' }
+
+setMethod("pivot",
+          signature(x = "GroupedData"),
+          function(x, colname, values=list()){
+              if (length(values) == 0) {
+                  result <- SparkR:::callJMethod(x@sgd, "pivot", colname)
+              }else {
```
Comment (shivaram) [inline R/pkg/R/group.R]: style nit: space after `}` here

File: R/pkg/NAMESPACE (hunk 2 above)
Comment (shivaram) [inline R/pkg/NAMESPACE]: +1

File: R/pkg/R/generics.R [hunk 8]
```diff
@@ -160,6 +160,10 @@ setGeneric("persist", function(x, newLevel) { standardGeneric("persist") })
 # @export
 setGeneric("pipeRDD", function(x, command, env = list()) { standardGeneric("pipeRDD")})
 
+# @rdname pivot
+# @export
+setGeneric("pivot", function(x, colname, values=list()) {  standardGeneric("pivot") })
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: style: `values = list()` with spaces

File: R/pkg/R/group.R [hunk 9]
```diff
@@ -126,6 +126,49 @@ methods <- c("avg", "max", "mean", "min", "sum")
...
+                  stopifnot(length(values) == length(unique(values)))
```
Comment (felixcheung) [inline R/pkg/R/group.R]: would be good to add a message - use `if () { stop() }` instead?

File: R/pkg/R/group.R [hunk 10]
```diff
@@ -126,6 +126,49 @@ methods <- c("avg", "max", "mean", "min", "sum")
...
+                  result <- SparkR:::callJMethod(x@sgd, "pivot", colname, values)
```
Comment (felixcheung) [inline R/pkg/R/group.R]: why do we need "SparkR:::" here and line 168?

Comment (felixcheung) [inline R/pkg/R/group.R]: and 163

**Comment by shivaram** [comment]
@mhnatiuk It looks like the tests are failing with the error message I've pasted below. From reading the code I think this is related to the `if` check we have in the pivot implementation where it should be `length(values) != length(unique(values))` instead of `==` as we have right now ?

```
1. Error: pivot GroupedData column ---------------------------------------------
error in evaluating the argument 'x' in selecting a method for function 'collect': 
  error in evaluating the argument 'x' in selecting a method for function 'summarize': 
  Values in list are not unique
Calls: pivot -> pivot
```

**Comment by shivaram** [comment]
@dongjoon-hyun @felixcheung -- I think @mhnatiuk is busy. If you have time it will be cool to submit another version of this PR as I think this is a useful function for R users.

**Comment by dongjoon-hyun** [comment]
This will be really useful. I'll make this PR up-to-date and adds credit description for @mhnatiuk .

**Comment by shivaram** [comment]
@mhnatiuk Given that #13786 was merged, can you close this PR ? Only the PR authors have permission to close a PR in the Spark project

**Comment by mhnatiuk** [comment]
I'll close that. Is it indicated that I did some work on that? Thanks.
MH

2016-06-21 7:46 GMT+02:00 Shivaram Venkataraman <notifications@github.com>:

**Comment by mhnatiuk** [comment]
[#13786](https://github.com/apache/spark/pull/13786) that finishes this work, was merged into branch-2.0. Closing.

**Pull request closed without merging**